[kenvynKwek] iP

kenvynKwek opened the pull request:
# DogeBot :dog2:
> "Im your best friend that helps you manage your tasks c:" -DogeBot

DogeBot remembers all the things you need to do with its **excellent** memory, DogeBot is:
- a dog
- ~~cute~~friendly
- _easy_ to use

All you need to do is:
1. access it from [here](https://github.com/kenvynKwek/ip)
2. run it
3. add your tasks and take your mind off things :)

Features:
- [x] Manage different types of tasks
- [x] Search keywords in tasks
- [ ] Graphical user interface (coming soon)

Just look at how ~~cute~~friendly DogeBot's response is when you `mark` a task:
```ruby
public static void mark(int index) throws DogeBotException {
        if (tasks.size() == 0) {
            throw new DogeBotException("Oops ! Try adding some tasks first c:");
        }

        tasks.get(index).markTask(true);
        System.out.println("Good job on completing a task ! You deserve a cookie C:");
        System.out.println("\t" + tasks.get(index).toString());
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

kenvynKwek committed:
Level-0: (a) renamed chatbot to 'DogeBot' (b) implemented initial skeleton message

kenvynKwek committed:
Level-1: Echo - echoes commands until 'bye'

kenvynKwek committed:
Level-2: Add, List - store tasks & output the list of tasks

kenvynKwek committed:
Level-3: Mark as Done - added mark task feature, created 'Task' class

kenvynKwek committed:
Level-4: ToDos, Events, Deadlines - added 'ToDos.java', 'Deadline.java', 'Event.java' and updated 'Duke.java' to take those 3 types of tasks

kenvynKwek committed:
dogebot logo rebrand (bc java can't compile certain ascii from prev logo)

kenvynKwek committed:
A-TextUiTesting: implemented test cases until Level-4 features

kenvynKwek committed:
'Duke.java' renamed to 'DogeBot.java', refactored code to switch cases (for throwing exceptions later)

kenvynKwek committed:
Level-5: Handle Errors: DogeBot.java: handles errors for task isBlank(), invalid input, mark/unmark inputMismatch, empty list

kenvynKwek committed:
DogeBot.java: changed 'tasks[100]' array to arraylist (for delete feature later)

kenvynKwek committed:
Level-6: Delete: implemented delete feature

kenvynKwek committed:
DogeBot.java: updated exceptions + updated automated text-ui-testing with new features

kenvynKwek committed:
Update try-catch exceptions

too many try-catch statements within switch-case

update code to have minimal try-catch statements

kenvynKwek committed:
Create 'tasklist.txt' for storing tasks to hard disk

The tasks are stored in an array.

Storing tasks in an array doesn't save the tasks when the program reruns.

Change the storing of tasks into a .txt file instead.

kenvynKwek committed:
Update bye() method and load 'tasklist.txt' to 'tasks' arraylist

Tasks are not stored in hard disk i.e 'tasklist.txt', stored in 'tasks' arraylist instead.

Task list progress is lost everytime the program is rerun.

Progress is loaded at the start into 'tasks' arraylist from 'tasklist.txt' and saved into 'tasklist.txt' when 'bye' command is entered.

kenvynKwek committed:
Merge branch 'branch-Level-7'

kenvynKwek committed:
Update deadline feature to store date/time

Previously stored deadline as string.

Deadline date/time now stored as LocalDateTime object.

kenvynKwek committed:
Update event feature to store date/time

Previously stores event duration as string.

Now stores event duration as LocalDateTime objects.

kenvynKwek committed:
Fix bug task done marker progress

Task done marker was saved into 'tasklist.txt' properly but is not loading correctly when the program starts.

Task done marker progress not tracked, shows wrong data/progress upon program start up.

Updated tasks constructors to initialise if the task has already been done, able to read progress correctly from 'tasklist.txt' now.

kenvynKwek committed:
Merge branch 'branch-Level-8'

kenvynKwek committed:
Refactor code and add 'Parser.java', 'Storage.java', 'TaskList.java', 'Ui.java'

Code was grouped together under 'DogeBot.java'.

Didn't make complete use of OOP principles and would be harder to implement new changes.

Refactored code to extract out closely related code as classes to make use of OOP principles under 'Parser.java', 'Storage.java', 'TaskList.java' and 'Ui.java'.

kenvynKwek committed:
Merge branch 'branch-A-MoreOOP'

mfjkri commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public String toString() {
+        return "E | " + super.toString() + " | " + start.toString() + " - " + end.toString();
```
Consider using String.format?

mfjkri commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,64 @@
+
+import java.io.FileWriter;
+import java.io.BufferedWriter;
+import java.io.IOException;
```
Could these import statements perhaps be reordered?

mfjkri commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,56 @@
+        boolean isLoop = true;
+
+        try {
+            switch (sc.next().toLowerCase()) {
```
I like the use of a switch statement here which makes the code readable and easy to follow!

mfjkri commented on src/main/java/DateTimeHandler.java:
```diff
@@ -0,0 +1,31 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+public class DateTimeHandler {
```
Consider writing descriptive header comments for all public classes/methods so that other developers will not be forced to read the code of the method to understand its exact behavior.

mfjkri reviewed (commented):
Overall clean and easy to read code. 

Do consider adding javadocs for all the public methods and classes.

kenvynKwek committed:
Add 'dogebot' and 'tasks' package

No packages previously.

Gets messy to find related .java files.

Grouped all files under 'dogebot' package and tasks type files under 'tasks' package, making related types easier to find.

kenvynKwek committed:
Merge branch 'branch-A-Packages'

mfjkri commented on src/main/java/dogebot/DogeBot.java:
```diff
@@ -0,0 +1,45 @@
+        tasks = new TaskList(storage.readFromTxtFile());
+    }
+
+    public void run() {
```
Consider adding javadocs here so that readers can easily understand the functionality of the method.

kenvynKwek committed:
Merge branch 'master' into add-gradle-support

kenvynKwek committed:
Implement gradle

Set up gradle to build and run DogeBot and minor edits to code for checkstyle.

kenvynKwek committed:
Merge branch 'branch-A-Gradle'

kenvynKwek committed:
Add JUnit tests

No JUnit tests previously.

Relied on text-ui-test runtest.sh test script, hard to maintain.

Added JUnit tests for non-trivial methods, easier to test/debug/maintain.

kenvynKwek committed:
Merge branch 'branch-A-JUnit'

kenvynKwek committed:
Create JAR file

DateTimeHandler.java uses DateTimeFormat code.

Works previously but creates problems when running from 'java -jar DogeBot.jar'.

Updated constructor's read-in-from-text-file's condition to initialise LocalDateTime object without using DateTimeFormat, 'java -jar DogeBot.jar' now runs without problems.

kenvynKwek committed:
Add JavaDoc

No header comments and Java Documention previously.

Hard for viewers to interpret the purpose of each method.

Java Documentation added, easier to understand the purpose of each method at a glance.

kenvynKwek committed:
Add checkstyle

To comply with coding standard for readability. No change in code.

kenvynKwek committed:
Add find feature

Implemented a find feature to search for tasks with input keyword. Updated 'TaskTest.java' to test new hasWord() method in 'Task.java'.

kenvynKwek committed:
Merge branch 'branch-A-JavaDoc'

kenvynKwek committed:
Merge branch 'branch-A-CodingStandard'

kenvynKwek committed:
Merge branch 'branch-Level-9'

elaineshijie commented on src/main/java/dogebot/DateTimeHandler.java:
```diff
@@ -0,0 +1,122 @@
+     */
+    public int monthToInt(String s) {
+        int month;
+        switch (s) {
```
Switch and case statements have the same indentation, good job!

elaineshijie commented on src/main/java/dogebot/TaskList.java:
```diff
@@ -0,0 +1,207 @@
+        }
+
+        tasks.get(index).markTask(true);
+        System.out.println("Good job on completing a task ! You deserve a cookie C:");
```
Would it be better to have methods in Ui to print these messages?

elaineshijie commented on src/main/java/dogebot/TaskList.java:
```diff
@@ -0,0 +1,207 @@
+     */
+    public static void find(String s) throws DogeBotException {
+        if (s.isBlank()) {
+            throw new DogeBotException("Oops ! The keyword cannot be empty :(");
```
Would it be better if exceptions are created based on the events eg. KeywordNotFoundException and these specific exceptions extends from DogeBotException?

elaineshijie reviewed (commented):
Code is generally well done but more can be done to organise code (Ui has methods to print output to user instead as Ui is a class to interact with user).

kenvynKwek committed:
Add GUI Part 3: Interacting with the user

Integrated JavaFX GUI to display user inputs into console and echo the command.

kenvynKwek committed:
Add GUI Part 4: Using FXML

Changed JavaFX to FXML and added intro message in GUI upon start up.

kenvynKwek committed:
Update GUI output

Previously still outputs to console and not GUI.

Refactored methods such that GUI is able to capture the output instead of outputting to console.

kenvynKwek committed:
Update GUI with exception handling

Updated methods to output exception messages to GUI instead of console.

kenvynKwek committed:
Merge branch 'branch-Level-10'

kenvynKwek committed:
Add assertions

User inputs for mark, unmark and delete feature only works under the assumption that the user inputs the task index correctly.

Mark, unmark and delete features might not work as intended.

Let's update the mark, unmark and delete features with assertions to test the correctness of the user input assumptions.

kenvynKwek committed:
Improve code quality

Code quality is intermediate.

Code quality can be improved for better readability.

Lets
* make the Happy Path Prominent in 'TaskList.java'
* apply SLAP in 'TaskList.java'
* structure code logically in 'Tasklist.java'

kenvynKwek committed:
Merge pull request #2 from kenvynKwek/branch-A-Assertions

Add assertions

kenvynKwek committed:
Merge branch 'master' into branch-A-CodeQuality

kenvynKwek committed:
Merge pull request #3 from kenvynKwek/branch-A-CodeQuality

Improve code quality

kenvynKwek committed:
Add reminder feature

Implemented reminder feature.

kenvynKwek committed:
Merge branch 'branch-BCD-Extension'

kenvynKwek committed:
BugFix: Tasks with reminder data load

Fixed bug to read tasks with reminder from .txt file

kenvynKwek committed:
GUI: increase application size

kenvynKwek committed:
Update storage path

Updated storage path so it works cross-platform.

kenvynKwek committed:
Ui.png

Screenshot to showcase my DogeBot in all its glory.

kenvynKwek committed:
Update README.md

kenvynKwek committed:
Add screenshots to assets

Add screenshots to docs/assets/images/ for the user guide.

kenvynKwek committed:
Update README.md with screenshots

kenvynKwek committed:
Fix README.md screenshots

kenvynKwek committed:
Update README.md with DogeBot description

kenvynKwek committed:
Update README.md

* List screenshot description added
* Bye screenshot description period removed
